gobblin-runtime/src/main/java/org/apache/gobblin/runtime/messaging/hdfs/FileSystemMessageBuffer.java

```diff
@@ -86,39 +103,77 @@ private void scheduleTaskForPollingNewMessages() {
+        if (fs.isFile(singlePath))
+        {
+          FSDataInputStream fis = fs.open(singlePath);
+          DynamicWorkUnitMessage message = DynamicWorkUnitSerde.deserialize(ByteStreams.toByteArray(fis));
+          messageList.add(message);
+          fs.delete(singlePath, true);
+          log.debug("Fetched message from HDFS path {}.", singlePath.toString());
+        }
+      }
+    } catch (IOException ioException) {
+      log.error("Failed to get message from HDFS path {}.", path.toString(), ioException);
+    }
+
+    return messageList;
+  }
+
+  private void persistMessage(byte[] serializedMsg, DynamicWorkUnitMessage message) {
+    try (FSDataOutputStream os = fs.create(this.getFilePathForMessage(message), true)) {
+      os.write(serializedMsg);
+      log.info("Persisted message into HDFS path {} for workunit {}.", path.toString(), message.getWorkUnitId());
+    } catch (IOException ioException) {
+      log.error("Failed to persist message into HDFS path {}.", path.toString(), ioException);
+    }
   }
 
-  private void persistMessage(byte[] serializeMsg) throws IOException {
-    // STUB: TODO GOBBLIN-1685 write serialized message to FileSystem
+  private Path getFilePathForMessage(DynamicWorkUnitMessage message) {
+    String messageFileName = message.getWorkUnitId() + "-" + System.currentTimeMillis();
+    Path messagePath = PathUtils.mergePaths(path, new Path(messageFileName));
+    return PathUtils.addExtension(messagePath, messageFileSuffix);
   }
 
   public static class Factory implements MessageBuffer.Factory<DynamicWorkUnitMessage> {
     private static volatile ScheduledExecutorService executorSingleton;
-    private Config cfg;
+    private final Config cfg;
+    private Path basePath;
+    private FileSystem fileSystem;
+    private long pollingRate;
     public Factory(Config cfg) {
-      // STUB: TODO GOBBLIN-1685 Add any necessary setup
       this.cfg = cfg;
+      this.setup();
     }
 
     @Override
     public MessageBuffer<DynamicWorkUnitMessage> getBuffer(String channelName) {
-      // STUB: TODO GOBBLIN-1685 Construct buffer object based on Typesafe Config
-      return null;
+      Path workunitPath = new Path(basePath, channelName);
+      log.info("HDFS directory for dynamic workunit is: " + workunitPath);
+
+      return new FileSystemMessageBuffer(fileSystem, channelName, pollingRate, workunitPath);
     }
 
-    private FileSystem getFileSystem(Config cfg) {
-      // STUB: TODO GOBBLIN-1685 Get filesystem based on current state
-      return null;
+    private void setup() {
+      String fsPathString = this.cfg.getString(DYNAMIC_WORKUNIT_HDFS_PATH);
+      this.basePath = new Path(fsPathString);
+      try {
+        this.fileSystem = this.basePath.getFileSystem(new Configuration());
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to detect workunit directory file system:", e);
+      }
+      pollingRate = this.cfg.getLong(DYNAMIC_WORKUNIT_HDFS_POLLING_RATE);
```

Comment:
Do we want a default value? Long term, we would make this feature opt-out instead of opt-in so I think we need a default value for usability.

Reply:
set 30000 millis